packages/perseus/src/components/math-input.jsx

```diff
@@ -268,6 +144,96 @@ class MathInput extends React.Component<$FlowFixMe, $FlowFixMe> {
         initialized = true;
     }
 
+    componentWillUnmount() {
+        window.removeEventListener("mousedown", this.handleMouseDown);
+        window.removeEventListener("mouseup", this.handleMouseUp);
+    }
+
+    // handlers:
+    // keep track of two related bits of state:
+    // * this.state.focused - whether the buttons are currently shown
+    // * this.mouseDown - whether a mouse click is active that started in the
+    //   buttons div
+
+    handleFocus: () => void = () => {
+        this.setState({focused: true});
+        // TODO(joel) fix properly - we should probably allow onFocus handlers
+        // to this property, but we need to work correctly with them.
+        // if (this.props.onFocus) {
+        //     this.props.onFocus();
+        // }
+    };
+
+    handleMouseDown: (MouseEvent) => void = (event) => {
+        // $FlowFixMe[incompatible-use]
+        // $FlowFixMe[incompatible-call]
+        const focused = ReactDOM.findDOMNode(this).contains(event.target);
+        this.mouseDown = focused;
+        if (!focused) {
+            this.setState({focused: false});
+        }
+    };
+
+    handleMouseUp: (MouseEvent) => void = () => {
+        // this mouse click started in the buttons div so we should focus the
+        // input
+        if (this.mouseDown) {
+            this.focus();
+        }
+        this.mouseDown = false;
+    };
+
+    handleBlur: (SyntheticFocusEvent<>) => void = (e) => {
+        // TODO(michaelpolyak): Consider trapping focus within the button group.
+        // Focusing back on the input when TAB out of the last button in the
+        // group. This will probably require ESCAPE key handling to enable to
+        // close (blur) the button group in order to focus on next page element.
+        if (
+            !this.mouseDown &&
+            // $FlowFixMe[incompatible-use]
+            // $FlowFixMe[incompatible-call]
+            !ReactDOM.findDOMNode(this).contains(e.relatedTarget)
+        ) {
+            this.setState({focused: false});
+        }
+    };
+
+    _shouldShowButtons: () => boolean = () => {
+        if (this.props.buttonsVisible === "always") {
+            return true;
+        }
+        if (this.props.buttonsVisible === "never") {
+            return false;
+        }
+        return this.state.focused;
+    };
+
+    insert: ($FlowFixMe) => void = (value) => {
+        const input = this.mathField();
+        if (_(value).isFunction()) {
+            value(input);
+        } else if (value[0] === "\\") {
+            input.cmd(value).focus();
+        } else {
+            input.write(value).focus();
+        }
+        input.focus();
+    };
```

Comment:
Oh snap.  That's a fancy function.  I wonder if you could type it as something like this: 
```suggestion
    insert: (value: (??Some sort of MathQuill object... or maybe $FlowFixMe here for now??) => void | string) => void = (value) => {
        const input = this.mathField();
        if (_(value).isFunction()) {
            value(input);
        } else if (value[0] === "\\") {
            input.cmd(value).focus();
        } else {
            input.write(value).focus();
        }
        input.focus();
    };
```

In fact, it almost looks like the parameter type here is the `Inserter` from `tex-buttons.jsx`. 

No action required if this is a rabbit hole. I'm fine with some $FlowFixMe's in code that we're working to type.  Some types are better than none!

Reply:
I gave this a shot, but couldn't figure out how to make Flow happy here:

```
        if (_(value).isFunction()) {
            value(input);
        }
```

Flow doesn't know it's a function.

Reply:
Ah.  It's hard to see how Flow will react when viewing the code on Github.  It was worth a try. 😄